hail/src/main/scala/is/hail/cxx/Emit.scala

```diff
@@ -701,26 +675,34 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int) {
         }
 
       case ir.MakeArray(args, _) =>
-        new ArrayEmitter("", "false", "", Some(args.length.toString)) {
+        val arrayRegion = EmitRegion.from(resultRegion, sameRegion)
+        val triplets = args.map { arg => outer.emit(arrayRegion, arg, env) }
+        new ArrayEmitter("", "false", "", Some(args.length.toString), arrayRegion) {
           def emit(f: (Code, Code) => Code): Code = {
             val sb = new ArrayBuilder[Code]
-            args.foreach { arg =>
-              val argt = outer.emit(arg, env)
-              sb += argt.setup
-              sb += f(argt.m, argt.v)
+            triplets.foreach { argt =>
+              sb +=
+                s"""
+                   |{
+                   |${ arrayRegion.defineIfUsed(sameRegion) }
```

Comment:
`arrayRegion.defineIfUsed` is called once per array element. Is that right? Aren't all the elements being built in the same region?

It also seems a bit dangerous to repeat the element processing code (`f(...)`) once per element. Maybe only unroll `MakeArray` if the size of `args` is small? This is outside the scope of the PR, so feel free to ignore.

Reply:
Oof, good catch. I've hoisted out the evaluation of `f(...)` since it should only be called once and affects whether or not we're defining the array region. I'm not super sure what we want to do with the efficiency of not repeating code---maybe at some point I'll generate function for `f` and call that one per element.

I think the region semantics for MakeArray should work exactly like the region semantics for all other stream producers---e.g. in ArrayRange, the region is declared inside the `for` loop, so it gets declared (if necessary) once for each element, and once you're processing the next element the old region is out of scope and so needs to be added to an in-scope region if we want to preserve the values. That's also why I added (what I think are) scoping brackets around each element of the `MakeArray`---it feels tidier to just think about it as "a single scope that handles processing of a single element of a stream, and the region is declared at the beginning and preserved as necessary".